firedrake/mesh.py

```diff
@@ -1537,6 +1686,175 @@ def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', kern
+    """
+    Create a vertex only mesh, immersed in a given mesh, with vertices
+    defined by a list of coordinates.
+
+    :arg mesh: The unstructured mesh in which to immerse the vertex only
+        mesh.
+    :arg vertexcoords: A list of coordinate tuples which defines the vertices.
+
+    .. note::
+
+        The vertex only mesh uses the same communicator as the input `mesh`.
+
+    .. note::
+
+        Meshes created from a coordinates `firedrake.Function` and immersed
+        manifold meshes are not yet supported.
+    """
+
+    import firedrake.functionspace as functionspace
+    import firedrake.function as function
+
+    mesh.init()
+
+    vertexcoords = np.asarray(vertexcoords, dtype=np.double)
+    gdim = mesh.geometric_dimension()
+    tdim = mesh.topological_dimension()
+    _, pdim = vertexcoords.shape
+
+    if isinstance(mesh.topology, ExtrudedMeshTopology):
+        raise NotImplementedError("Extruded meshes are not supported")
+
+    if gdim != tdim:
+        raise NotImplementedError("Immersed manifold meshes are not supported")
+
+    if mesh.coordinates.function_space().ufl_element().degree() > 1:
+        raise NotImplementedError("Only straight edged meshes are supported")
+
+    if hasattr(mesh, "_made_from_coordinates") and mesh._made_from_coordinates:
+        raise NotImplementedError("Meshes made from coordinate fields are not yet supported")
+
+    if pdim != tdim:
+        raise ValueError(f"Mesh topological dimension {tdim} must match point list dimension {pdim}")
+
+    swarm = _pic_swarm_in_plex(mesh.topology._plex, vertexcoords, fields=[("parentcellnum", 1, IntType)])
+
+    dmswarm.label_pic_parent_cell_nums(swarm, mesh)
+
+    # Topology
+    topology = VertexOnlyMeshTopology(swarm, mesh.topology, name="swarmmesh", reorder=False)
+
+    # Geometry
+    tcell = topology.ufl_cell()
+    cell = tcell.reconstruct(geometric_dimension=gdim)
+    element = ufl.VectorElement("DG", cell, 0)
+    # Create mesh object
+    vmesh = MeshGeometry.__new__(MeshGeometry, element)
+    vmesh._topology = topology
+    vmesh._parent_mesh = mesh
+
+    def callback(self):
+        """Finish initialisation."""
+        del self._callback
+        # Finish the initialisation of mesh topology
+        self.topology.init()
+
+        coordinates_fs = functionspace.VectorFunctionSpace(self.topology, "DG", 0,
+                                                           dim=gdim)
+
+        coordinates_data = dmswarm.reordered_coords(swarm, coordinates_fs.dm.getDefaultSection(),
+                                                    (self.num_vertices(), gdim))
+
+        coordinates = function.CoordinatelessFunction(coordinates_fs,
+                                                      val=coordinates_data,
+                                                      name="Coordinates")
+
+        self.__init__(coordinates)
+
+    vmesh._callback = callback
+    return vmesh
+
+
+def _pic_swarm_in_plex(plex, coords, **kwargs):
```

Comment:
If you only accept `fields` as a kwarg, write this as:

```
def _pic_swarm_in_plex(plex, coords, *, fields=())
```

Now no need to do the get with default later.

Reply:
Thanks, done.